File: ruby/sieve/sieve.rb
require 'set'

# Sieve of Eratosthenes
class Sieve
  def initialize(max)
    @max = max
  end

  def primes
    prime_numbers = Set.new(2..@max)

    2.upto(Math.sqrt(@max).ceil) do |x|
      ((x + x)..@max).step(x) do |np|
        prime_numbers.delete(np)
      end
    end

    prime_numbers.to_a
  end
end

File: elixir/sieve/sieve.exs
defmodule Sieve do

  @doc """
  Generates a list of primes up to a given limit.
  """
  @spec primes_to(non_neg_integer) :: [non_neg_integer]
  def primes_to(limit) do
    max = :math.sqrt(limit)
          |> Float.ceil
          |> round

    set = MapSet.new(2..limit)

    primes(2, max, limit, set) |> MapSet.to_list |> Enum.sort
  end

  defp primes(start, max, limit, set) when start > max, do: set
  defp primes(start, max, limit, set) do
    to_remove = start..limit
                |> Enum.take_every(start)
                |> Enum.drop(1)
                |> MapSet.new

    primes(start + 1, max, limit, MapSet.difference(set, to_remove))
  end
end
